pymongo/_asyncio_task.py

```diff
@@ -0,0 +1,40 @@
+# Copyright 2024-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A custom asyncio.Task that allows checking if a task has been sent a cancellation request.
+Can be removed once we drop Python 3.10 support in favor of asyncio.Task.cancelling."""
+
+
+from __future__ import annotations
+
+import asyncio
+from typing import Any, Coroutine, Optional
+
+
+class _Task(asyncio.Task):
+    def __init__(self, coro: Coroutine[Any, Any, Any], *, name: Optional[str] = None) -> None:
+        super().__init__(coro, name=name)
+        self._cancelled: bool = False
+        asyncio._register_task(self)
```

Comment:
Oh I see the problem. That wouldn't work because asyncio.all_tasks() would return the non-wrapper class.

Reply:
The problem still remains though. If we can't do this using only public apis then we can't add this workaround at all.

Reply:
`asyncio._register_task` is documented in the asyncio docs here: https://docs.python.org/3/library/asyncio-extending.html as the only way to extend `Task` functionality.

Reply:
Oh I see, I didn't realize it was a documented api. Still, I'm not a big fan of this because the docs discourage what we're doing here:
> Use the high-level [asyncio.create_task()](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task) function to create Tasks, or the low-level [loop.create_task()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_task) or [ensure_future()](https://docs.python.org/3/library/asyncio-future.html#asyncio.ensure_future) functions. Manual instantiation of Tasks is discouraged.

Reply:
I agree it's not ideal, but it's very simple. 

Another possible solution would be to override the `asyncio.Task` import to use a vendored implementation of the 3.13 `Task`, but then we're mucking around with imported libraries. We could also mutate `_Task.task` to have a `cancelled` attribute and monkey-patch its `cancel` method.

Reply:
The above passes all tests for me locally and doesn't use any private `asyncio` methods or import hacking, only `MethodType` tricks.

Reply:
If we own the task and its cancellation, can't we keep a mapping of tasks to cancelled state ourselves?

Reply:
We don't own the cancellation when the test runner cancels the task at the end of the test that created it.

Reply:
Okay I relent for now since we plan remove this code anyway once we figure out the real issue causing the hangs.

Reply:
I'll add a TODO for us to come back and revisit this at that time.